contracts/erc721/ERC721PreApproval.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2021 the ethier authors (github.com/divergencetech/ethier)
+pragma solidity >=0.8.0 <0.9.0;
+
+import "../thirdparty/opensea/OpenSeaGasFreeListing.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+/// @notice Pre-approval of OpenSea proxies for gas-less listing
+/// @dev This wrapper allows users to revoke the pre-approval of their
+/// associated proxy and emits the corresponding events. This is necessary for
+/// external tools to index approvals correctly and inform the user.
+/// @dev The pre-approval is triggered on a per-wallet basis during the first
+/// transfer transactions. It will only be enabled for wallets with an existing
+/// proxy. Not having a proxy incurs a gas overhead.
+/// @dev This wrapper optimizes for the following scenario:
+/// - The majority of users already have a wyvern proxy
+/// - Most of them want to transfer tokens via wyvern exchanges
+abstract contract ERC721PreApproval is ERC721 {
```

Comment:
If it's `is IERC721` then anyone can use this with arbitrary `ERC721` implementations, not just OZ.

Reply:
Not quite possible here, because I rely on OZ's callbacks like `_beforeTokenTransfer` and `balanceOf`. Maybe we could introduce a basic internal ERC721 layout (with custom callbacks) for us at some point?